File: app/src/main/java/com/loopeer/codereader/api/ApiService.java
package com.loopeer.codereader.api;

import android.app.Application;

import com.loopeer.codereader.CodeReaderApplication;

import java.io.File;
import java.util.MissingResourceException;

import okhttp3.Cache;
import okhttp3.OkHttpClient;
import retrofit2.Retrofit;
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory;
import retrofit2.converter.gson.GsonConverterFactory;

public class ApiService {

    public static final String API_URL = "https://github.com/";

    private static ApiService sInstance;

    private Retrofit mRetrofit;

    public static synchronized ApiService getInstance() {
        if (sInstance == null) {
            sInstance = new ApiService();
        }
        return sInstance;
    }


    private OkHttpClient getClient() {
        return createOkHttpClient(CodeReaderApplication.getInstance());
    }

    static final int DISK_CACHE_SIZE = 50 * 1024 * 1024; // 50MB

    static OkHttpClient createOkHttpClient(Application app) {
        OkHttpClient.Builder httpClient = new OkHttpClient.Builder();
        File cacheDir = new File(app.getCacheDir(), "http");
        Cache cache = new Cache(cacheDir, DISK_CACHE_SIZE);
        httpClient.cache(cache);
        return httpClient.build();
    }

    protected Retrofit.Builder newRestAdapterBuilder() {
        return new Retrofit.Builder();
    }

    protected Retrofit getRetrofit() {
        if (mRetrofit == null) {
            try {
                mRetrofit = newRestAdapterBuilder()
                        .client(getClient())
                        .addConverterFactory(GsonConverterFactory.create())
                        .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                        .baseUrl(API_URL)
                        .build();
            } catch (NullPointerException e) {
                throw new MissingResourceException("Define your endpoint in api_url string resource.", getClass().getName(), "api_url");
            }
        }

        return mRetrofit;
    }

    public static <T> T create(Class<T> service) {
        return getInstance().getRetrofit().create(service);
    }

}

File: code-reader-kt/src/main/java/com/loopeer/codereaderkt/api/ApiService.kt
package com.loopeer.codereaderkt.api

import android.app.Application
import com.loopeer.codereaderkt.BuildConfig
import com.loopeer.codereaderkt.CodeReaderApplication
import okhttp3.Cache
import okhttp3.OkHttpClient
import retrofit2.Retrofit
import retrofit2.adapter.rxjava.RxJavaCallAdapterFactory
import retrofit2.converter.gson.GsonConverterFactory
import java.io.File
import java.util.*
import java.util.concurrent.TimeUnit


open class ApiService {

    private var mRetrofit: Retrofit? = null


    private fun getClient(): OkHttpClient {
        return createOkHttpClient(CodeReaderApplication.instance)
    }

    val retrofit: Retrofit
        get() {
            if (mRetrofit == null) {
                try {
                    mRetrofit = Retrofit.Builder()
                            .client(getClient())
                            .addConverterFactory(GsonConverterFactory.create())
                            .addCallAdapterFactory(RxJavaCallAdapterFactory.create())
                            .baseUrl(API_URL)
                            .build()
                } catch (e: NullPointerException) {
                    throw MissingResourceException("Define your endpoint in api_url string resource.", javaClass.name, "api_url")
                }

            }

            return this.mRetrofit!!
        }

    companion object {

        val API_URL = "https://api.github.com/"

        private var sInstance: ApiService? = null

        val instance: ApiService
            @Synchronized get() {
                if (sInstance == null) {
                    sInstance = ApiService()
                }
                return sInstance!!
            }

        private val DISK_CACHE_SIZE = 50 * 1024 * 1024 // 50MB

        internal fun createOkHttpClient(app: Application): OkHttpClient {
            val httpClient = OkHttpClient.Builder()

            if (BuildConfig.DEBUG) {
                val loggingInterceptor = HttpJsonLoggingInterceptor()
                loggingInterceptor.level = HttpJsonLoggingInterceptor.Level.BODY
                httpClient.addInterceptor(loggingInterceptor)
            }

            httpClient.connectTimeout(1, TimeUnit.HOURS) // connect timeout
            httpClient.readTimeout(1, TimeUnit.HOURS)
            val cacheDir = File(app.cacheDir, "http")
            val cache = Cache(cacheDir, DISK_CACHE_SIZE.toLong())
            httpClient.cache(cache)
            return httpClient.build()
        }

        fun <T> create(service: Class<T>): T = instance.retrofit.create(service)
    }

}